crates/circuit/src/slice.rs

```diff
@@ -80,6 +80,15 @@ impl<'py> PySequenceIndex<'py> {
             }
         }
     }
+
+    /// Given an integer (which may be negative) get a valid unsigned index for a sequence.
+    pub fn convert_idx(index: isize, length: usize) -> Result<usize, PySequenceIndexError> {
+        let wrapped_index = match PySequenceIndex::Int(index).with_len(length)? {
+            SequenceIndex::Int(result) => result,
+            _ => unreachable!(), // we will always match on Int(..)
+        };
+        Ok(wrapped_index)
+    }
```

Comment:
It might be marginally cleaner to pull out this logic from `with_len`, and have `with_len` call this instead.

Reply:
Ah yes, then we can avoid the `unreachable!` branch, good idea! Done in f9c114c.